File: npmjs/pray-cli/src/cli/commands/update-latest.ts
import { existsSync, readFileSync } from "node:fs";
import {
  latestConstraintForPackage,
  versionSatisfies,
} from "../../constraint.js";
import { PrayError } from "../../errors.js";
import { readLockfile } from "../../lockfile/index.js";
import { parseManifest } from "../../manifest/index.js";
import { replacePackageDeclaration } from "../../manifest/package-declaration.js";
import { resolveProject } from "../../resolve/project.js";
import {
  applyPathUpstreamLatestConstraints,
  type PathUpstreamLatestConstraint,
  planPathUpstreamLatestConstraints,
} from "../../resolve/upstream-latest.js";
import { applyPathUpstreamRefreshes } from "../../resolve/upstream-refresh.js";
import {
  lockfilePath,
  manifestPath,
  resolveCurrentProject,
} from "../invocation.js";
import { updateResolveOptions, writeUpdate } from "./update-core.js";

export async function updateLatestCommand(
  packageName: string | undefined,
  json: boolean,
  dryRun = false,
): Promise<void> {
  const path = manifestPath();
  const originalText = readFileSync(path, "utf8");
  let manifestText = originalText;
  const options = updateResolveOptions(packageName);
  const project = await resolveCurrentProject(options);
  if (
    packageName &&
    !project.manifest.packages.some((entry) => entry.name === packageName)
  ) {
    throw PrayError.manifest(`package ${packageName} not found`);
  }

  const manifestUpdates: Array<{
    name: string;
    from_constraint: string;
    to_constraint: string;
    registry_latest_version: string;
  }> = [];

  for (const resolved of project.packages) {
    if (packageName && resolved.declaration.name !== packageName) {
      continue;
    }
    const registryLatest = resolved.registryLatestVersion;
    if (!registryLatest) {
      continue;
    }
    if (versionSatisfies(registryLatest, resolved.declaration.constraint)) {
      continue;
    }
    const newConstraint = latestConstraintForPackage(
      resolved.declaration.constraint,
      registryLatest,
    );
    if (!versionSatisfies(registryLatest, newConstraint)) {
      throw PrayError.resolution(
        `derived constraint ${newConstraint} does not admit registry latest ${registryLatest} for ${resolved.declaration.name}`,
      );
    }
    manifestUpdates.push({
      name: resolved.declaration.name,
      from_constraint: resolved.declaration.constraint,
      to_constraint: newConstraint,
      registry_latest_version: registryLatest,
    });
    manifestText = replacePackageDeclaration(manifestText, {
      ...resolved.declaration,
      constraint: newConstraint,
    });
  }

  const previous = existsSync(lockfilePath())
    ? readLockfile(lockfilePath())
    : undefined;
  const upstreamPlans = await planPathUpstreamLatestConstraints(
    project,
    previous,
    packageName,
    options,
  );
  const upstreamConstraintUpdates = upstreamPlans.map((plan) => ({
    name: plan.packageName,
    from_constraint: plan.currentConstraint,
    to_constraint: plan.newConstraint,
    latest_version: plan.latestVersion,
  }));

  if (!json) {
    if (manifestUpdates.length === 0 && upstreamPlans.length === 0) {
      process.stdout.write(
        "All package constraints already allow latest versions\n",
      );
    } else {
      for (const update of manifestUpdates) {
        process.stdout.write(
          `Prayfile: ${update.name} constraint ${update.from_constraint} -> ${update.to_constraint} (registry latest ${update.registry_latest_version})\n`,
        );
      }
      printLatestUpstreamConstraints(upstreamPlans);
    }
  }

  let candidate = project;
  if (manifestUpdates.length > 0 || upstreamPlans.length > 0) {
    if (!dryRun) {
      applyPathUpstreamLatestConstraints(upstreamPlans);
    }
    candidate = await resolveProject(
      path,
      options,
      parseManifest(manifestText),
    );
    if (
      !dryRun &&
      (await applyPathUpstreamRefreshes(
        candidate,
        previous,
        packageName,
        options,
      ))
    ) {
      candidate = await resolveProject(
        path,
        options,
        parseManifest(manifestText),
      );
    }
  } else if (
    !dryRun &&
    (await applyPathUpstreamRefreshes(project, previous, packageName, options))
  ) {
    candidate = await resolveProject(
      path,
      options,
      parseManifest(manifestText),
    );
  }
  await writeUpdate(
    candidate,
    packageName,
    json,
    manifestUpdates,
    manifestText === originalText ? undefined : manifestText,
    dryRun,
    upstreamConstraintUpdates,
  );
}

function printLatestUpstreamConstraints(
  plans: readonly PathUpstreamLatestConstraint[],
): void {
  for (const plan of plans) {
    process.stdout.write(
      `${plan.packageName} upstream ${plan.currentConstraint} -> ${plan.newConstraint} (latest ${plan.latestVersion})\n`,
    );
  }
}

File: rubygems/pray-cli/lib/pray/cli/commands/update_latest.rb
# frozen_string_literal: true

module Pray
  module CLI
    def update_latest_command(package, dry_run:, offline:)
      path = manifest_path
      original_text = Pray.read_manifest_text(path)
      manifest_text = original_text
      preview_options = ResolveOptions.new(
        offline: offline,
        refresh: true,
        refresh_source_revisions: true,
        ignore_locked_versions: true
      )
      project = resolve_current_project(preview_options)
      if package && project.manifest.packages.none? { |entry| entry.name == package }
        raise Error.manifest("package #{package} not found")
      end

      manifest_updates = []
      project.packages.each do |resolved|
        next if package && resolved.declaration.name != package
        next unless resolved.registry_latest_version
        next if Constraint.version_satisfies(resolved.registry_latest_version, resolved.declaration.constraint)

        new_constraint = Constraint.latest_constraint_for_package(
          resolved.declaration.constraint,
          resolved.registry_latest_version
        )
        unless Constraint.version_satisfies(resolved.registry_latest_version, new_constraint)
          raise Error.resolution(
            "derived constraint #{new_constraint} does not admit registry latest #{resolved.registry_latest_version} for #{resolved.declaration.name}"
          )
        end

        manifest_updates << {
          name: resolved.declaration.name,
          from_constraint: resolved.declaration.constraint,
          to_constraint: new_constraint,
          registry_latest_version: resolved.registry_latest_version
        }
        updated = resolved.declaration.dup
        updated.constraint = new_constraint
        manifest_text = Pray.replace_package_declaration(manifest_text, updated)
      end

      previous = File.exist?(lockfile_path) ? Pray.read_lockfile(lockfile_path) : nil
      upstream_plans = Upstream.plan_path_upstream_latest_constraints(
        project, previous, package, preview_options
      )
      print_latest_constraint_plans(manifest_updates, upstream_plans)

      return if dry_run

      Upstream.apply_path_upstream_latest_constraints(upstream_plans)
      Transaction.write_file(path, manifest_text) if manifest_text != original_text
      unlocked = package ? Set[package] : Set.new
      options = ResolveOptions.new(
        offline: offline,
        refresh: true,
        refresh_source_revisions: true,
        ignore_locked_versions: package.nil?,
        unlocked_packages: unlocked
      )
      current = resolve_current_project(options)
      Upstream.apply_path_upstream_refreshes(current, previous, package, options)
      install_command(
        {
          locked: false,
          frozen: false,
          offline: offline,
          refresh: true,
          ignore_locked_versions: package.nil?,
          unlocked_packages: unlocked
        }
      )
    end

    def print_latest_constraint_plans(manifest_updates, upstream_plans)
      if manifest_updates.empty? && upstream_plans.empty?
        puts "All package constraints already allow latest versions"
        return
      end

      manifest_updates.each do |update|
        puts "Prayfile: #{update[:name]} constraint #{update[:from_constraint]} -> #{update[:to_constraint]} " \
          "(registry latest #{update[:registry_latest_version]})"
      end
      upstream_plans.each do |plan|
        puts "#{plan.package_name} upstream #{plan.current_constraint} -> #{plan.new_constraint} (latest #{plan.latest_version})"
      end
    end
  end
end
